packages/web/src/components/buy-sell-modal/hooks/useTokenSwapForm.ts

```diff
@@ -0,0 +1,188 @@
+  }, [getInputBalance, inputToken.balance])
+
+  // Use Jupiter API to get real-time exchange rate
+  const {
+    data: exchangeRateData,
+    isLoading: isExchangeRateLoading,
+    error: exchangeRateError
+  } = useTokenExchangeRate({
+    inputTokenSymbol,
+    outputTokenSymbol,
+    inputAmount: numericInputAmount > 0 ? numericInputAmount : 1
+  })
+
+  // Calculate the output amount based on the exchange rate
+  const outputAmount = useMemo(() => {
+    if (numericInputAmount <= 0) return 0
+    if (isExchangeRateLoading || !exchangeRateData) return 0
+
+    return exchangeRateData.rate * numericInputAmount
+  }, [numericInputAmount, exchangeRateData, isExchangeRateLoading])
+
+  // Validate the input amount
+  useEffect(() => {
+    if (numericInputAmount === 0) {
+      setError(null)
+      return
+    }
+
+    if (numericInputAmount < min) {
+      setError(`Minimum amount is ${min} ${inputToken.symbol}`)
+      return
+    }
+
+    if (numericInputAmount > max) {
+      setError(`Maximum amount is ${max} ${inputToken.symbol}`)
+      return
+    }
+
+    // Check if user has enough balance
+    const balance = getInputBalance()
+    if (balance !== undefined && numericInputAmount > balance) {
+      setError(formatBalanceError(numericInputAmount))
+      return
+    }
+
+    setError(null)
+  }, [
+    numericInputAmount,
+    min,
+    max,
+    getInputBalance,
+    formatBalanceError,
+    inputToken.symbol
+  ])
+
+  // Update the parent component with transaction data
+  useEffect(() => {
+    if (onTransactionDataChange) {
+      onTransactionDataChange({
+        inputAmount: numericInputAmount,
+        outputAmount,
+        isValid: numericInputAmount > 0 && !error && !isExchangeRateLoading
+      })
+    }
+  }, [
+    numericInputAmount,
+    outputAmount,
+    error,
+    isExchangeRateLoading,
+    onTransactionDataChange
+  ])
+
+  // Handle input changes
+  const handleInputAmountChange = useCallback((value: string) => {
+    // Allow only valid number input
+    if (value === '' || /^\d*\.?\d*$/.test(value)) {
+      setInputAmount(value)
+    }
+  }, [])
+
+  // Handle max button click
+  const handleMaxClick = useCallback(() => {
+    const balance = getInputBalance()
+    if (balance !== undefined) {
+      // Limit to MAX_AMOUNT
+      const finalAmount = Math.min(balance, max)
+      setInputAmount(finalAmount.toString())
+    }
+  }, [getInputBalance, max])
+
+  // Use the real exchange rate if available, otherwise null (no default fallback)
+  const currentExchangeRate = exchangeRateData ? exchangeRateData.rate : null
+
+  return {
+    inputAmount,
+    numericInputAmount,
+    outputAmount,
+    error,
+    exchangeRateError,
+    isExchangeRateLoading,
+    isBalanceLoading,
+    availableBalance,
+    currentExchangeRate,
+    handleInputAmountChange,
+    handleMaxClick,
+    // Including these for the component that consumes this hook
+    inputToken,
+    outputToken
+  }
```

Comment:
I generally like this hook but did we consider formik + zod schemas?

Reply:
yea using it in my local, gonna close this PR out and just continue on with big PR